### PR TITLE
Add matcher to detect softmax op and add pass to raise op

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
Skimmed through this. Looks good to me, thanks for kicking these off. Ill let someone more knowledgeable about matcher address any specific issues, but that reads fine to me as well.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp]: ah I finally understood how to get the captured data, my bad I had meant to ask how to do that and forgot I had this hack.

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp
Comment (ThomasRaoux) [reply llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp]: I don't see anything. I moved the code into an helper locally for now

File: llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h
Comment (ftynse) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h:478]: Naming nit: canonical doesn't have a trailing e in English :)

**Review by ftynse — APPROVED** [review]

**Comment by ftynse** [comment]
LGTM as long as code owners have no objections.

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
The matching implementation makes sense to me.

However the raising aspects trigger some red flags for me .. 
I have the impression that the raising aspect may be trying to work around difficulties with the grouping at flow + delayed fusion implementation.

If so, are we starting to leak codegen constraints to the frontend opset and does this generalize going forward?

Not questioning the value of having more and better named ops that simplify parts of the compiler and make the IR significantly nicer to operate on.

In particular, one of the key value of such named ops is "idiomatic abstraction that avoids needing complex matchers". Yet we are precisely building very complex one-off matchers.
It also seems somewhat brittle .. what if the matched constant is different for some reason: are we going to see a 100x perf drop because we special cased a codegen path that we are not able to match? 

To be clear, I think there is room for such a "raising" transformation in general: this is the inverse of the Generalize transformation for one particular use case.
It is maybe also one of the evolutions of matcher/generator infra to be able to save us from the hardships of writing complex matchers as an inverse problem.

Still, this smells very fishy and I am wondering whether the implications have been well discussed: are we going to latch on this op to perform the equivalent of fusion during codegen?

Putting a blocker on this so this can be discussed.

**Comment by ftynse** [comment]
Fundamentally, the result of _any_ matcher capturing more than a single operation can be materialized in the payload IR as a new operation. At the same time, it does not necessarily imply fusion, it may be a just a mechanism to preserve the result of the match through passes and then decompose it back into constituents that will be fed to downstream transformations without any fusion.

One possibility provided by matchers is indeed capturing a narrowly specific case and using a customized implementation. I am seeing it as a necessary escape hatch that at least leverages the common infrastructure as opposed to rolling a custom detection-and-transformation pass.

**Comment by MaheshRavishankar** [comment]
> The matching implementation makes sense to me.
> 
> However the raising aspects trigger some red flags for me .. I have the impression that the raising aspect may be trying to work around difficulties with the grouping at flow + delayed fusion implementation.
> 
> If so, are we starting to leak codegen constraints to the frontend opset and does this generalize going forward?
> 

The issue here is what are the codegen constraints. That can only be defined if we know a generic mechanism for generating the executable code. There is one generic characterization that I can think of, but some ops dont fit into those. For example, in the case of softmax, there is a fairly ad-hoc rematerialization of some ops done to actually generate the final code, i.e. the strategy used for generating code for softmax does not fit into how everything else is handled. Instead of relying on some ad-hoc transformations to generate code for these ops, that have to thread a narrow path to hit exactly the sequence of ops that sequence of transformations was meant for, matching it and replacing it with an op, then lets you use the op name to steer code-generation. Is it sustainable, thats TBD. Maybe as more operations of these kinds are handled, there might be a generalization that  covers all of these. I dont see a reason to block this PR for that reason.

> Not questioning the value of having more and better named ops that simplify parts of the compiler and make the IR significantly nicer to operate on.
> 
> In particular, one of the key value of such named ops is "idiomatic abstraction that avoids needing complex matchers". Yet we are precisely building very complex one-off matchers. It also seems somewhat brittle .. what if the matched constant is different for some reason: are we going to see a 100x perf drop because we special cased a codegen path that we are not able to match?

No, it wont be 100x. If you decompose softmax it will become two dispatches instead of 1. That wont be 100x off (unless there is a backend issue where it cant handle reduction for whatever reason. It wont be "the best possible way", but should still be reasonable.

> 
> To be clear, I think there is room for such a "raising" transformation in general: this is the inverse of the Generalize transformation for one particular use case. It is maybe also one of the evolutions of matcher/generator infra to be able to save us from the hardships of writing complex matchers as an inverse problem.

Yup agreed to that... having better matching infra (what is used today is along that path IMO) would make this easier.

> 
> Still, this smells very fishy and I am wondering whether the implications have been well discussed: are we going to latch on this op to perform the equivalent of fusion during codegen?
> 
> Putting a blocker on this so this can be discussed.

Dont see a reason to put a blocker on this. There are so many PRs that have landed that could have been blocked the same way. We can discuss this after the fact.

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by nicolasvasilache** [comment]
> Dont see a reason to put a blocker on this. There are so many PRs that have landed that could have been blocked the same way. We can discuss this after the fact.

Discussion has occurred, releasing.

**Review by nicolasvasilache — APPROVED** [review]

**Comment by nicolasvasilache** [comment]
> No, it wont be 100x. If you decompose softmax it will become two dispatches instead of 1. That wont be 100x off (unless there is a backend issue where it cant handle reduction for whatever reason. It wont be "the best possible way", but should still be reasonable.

Great, this is the high order bit I wanted to see, thanks!

**Pull request merged**
